### PR TITLE
Arbitrary pulumi commands

### PR DESCRIPTION
**Pull request by JimMadge**
## :white_check_mark: Checklist

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).
- [x] You have marked this pull request as a **draft** and added `'[WIP]'` to the title if needed (if you're not yet ready to merge).

### :arrow_heading_up: Summary

Add command for running Pulumi CLI commands using a DSH project+stack

### :microscope: Tests

Tests included, but also locally

```
dsh pulumi run shm 'stack ls'
dsh pulumi run shm 'stack output'
dsh pulumi run shm 'import ...'
```

### CODE REVIEW

**Review by craddm — APPROVED** [review]
LGTM, tested it too

File: data_safe_haven/commands/cli.py [hunk 1]
```diff
@@ -76,6 +77,9 @@ def callback(
 application.add_typer(
     context_command_group, name="context", help="Manage Data Safe Haven contexts."
 )
+application.add_typer(
+    pulumi_command_group, name="pulumi", help="Interact with the Pulumi CLI."
```
Comment (jemrobinson) [inline data_safe_haven/commands/cli.py]:
```suggestion
    pulumi_command_group, name="pulumi", help="(Advanced) interact directly with the Pulumi CLI."
```

File: data_safe_haven/commands/pulumi.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    SHM = auto()
+    SRE = auto()
+
+
+@pulumi_command_group.command()
+def run(
+    project_type: Annotated[
+        ProjectType,
+        typer.Argument(
+            # metavar="{SHM|SRE}",
```
Comment (jemrobinson) [inline data_safe_haven/commands/pulumi.py]: Do we need this comment?

**Review by jemrobinson — CHANGES_REQUESTED** [review]

File: data_safe_haven/commands/pulumi.py (hunk 2 above)
Comment (JimMadge) [reply data_safe_haven/commands/pulumi.py]: No, I was trying to make the help text clearer, but it didn't work as I expected it to. Possibly a conflict with the rich rendering. In any case it isn't needed now.

File: tests/conftest.py [hunk 3]
```diff
@@ -237,3 +244,56 @@ def mock_get_keyvault_key(self, key_name, key_vault_name):  # noqa: ARG001
 #         return [MockStorageAccountKey("key1"), MockStorageAccountKey("key2")]
 
 #     monkeypatch.setattr(AzureApi, "get_storage_account_keys", mock_get_storage_account_keys)
+
```
Comment (jemrobinson) [inline tests/conftest.py]: I know it's not part of this PR, but can we clean up the commented-out lines 237-246 here?

**Review by jemrobinson — APPROVED** [review]
One minor comment but overall I'm happy with this.

The indirection of `hatch run dsh pulumi run shm "refresh"` is really something though!

File: data_safe_haven/infrastructure/project_manager.py [hunk 4]
```diff
@@ -346,6 +346,15 @@ def refresh(self) -> None:
             msg = f"Pulumi refresh failed.\n{exc}"
             raise DataSafeHavenPulumiError(msg) from exc
 
+    def run_pulumi_command(self, command: str) -> str:
+        """Run a Pulumi CLI command using this project and stack."""
+        try:
+            result = self.stack._run_pulumi_cmd_sync(command.split())
```
Comment (jemrobinson) [inline data_safe_haven/infrastructure/project_manager.py:352]: Note that interactive commands won't work as we aren't passing stdin to the Pulumi process.

Comment (JimMadge) [reply data_safe_haven/infrastructure/project_manager.py:352]: :+1: I've added to the docstring.

**Pull request merged**